samples/2_Cookbook/13_occupancy/occupancy.cpp

```
@@ -0,0 +1,167 @@
+    }
+}
+
+void launchKernel(float* C, float* A, float* B, int N, bool manual){
+     
+     hipDeviceProp_t devProp;
+     HIP_CHECK(hipGetDeviceProperties(&devProp, 0));
+
+     hipEvent_t start, stop;
+     HIP_CHECK(hipEventCreate(&start));
+     HIP_CHECK(hipEventCreate(&stop));
+     float eventMs = 1.0f;
+     int activeWarps, maxWarps;
+      
+     if (manual){
+	blockSize = threadsperblock; 
+     	gridSize  = blocks;
+        std::cout << std::endl << "Manual Configuration with block size " << blockSize << std::endl;
+     }
+     else{
+     	hipOccupancyMaxPotentialBlockSize(&gridSize, &blockSize, multiply, 0, 0);
+	std::cout << std::endl << "Automatic Configuation based on hipOccupancyMaxPotentialBlockSize " << std::endl;
+	std::cout << "Suggested blocksize is " << blockSize << ", Minimum gridsize is " << gridSize << std::endl; 
+     }
+
+     // Record the start event
+     HIP_CHECK(hipEventRecord(start, NULL));  
+
+     // Launching the Kernel from Host
+     hipLaunchKernelGGL(multiply, dim3(gridSize), dim3(blockSize), 0, 0, C, A, B, NUM);
+     
+     // Record the stop event
+     HIP_CHECK(hipEventRecord(stop, NULL));
+     HIP_CHECK(hipEventSynchronize(stop));
+         
+     HIP_CHECK(hipEventElapsedTime(&eventMs, start, stop));
+     printf("kernel Execution time = %6.3fms\n", eventMs);
+
+     //Calculate Occupancy
+     uint32_t numBlock = 0;
+     hipOccupancyMaxActiveBlocksPerMultiprocessor(&numBlock, multiply, blockSize, 0);
+     
+     if(devProp.warpSize){
+	activeWarps = numBlock* blockSize/devProp.warpSize;
+	maxWarps = devProp.maxThreadsPerMultiProcessor/devProp.warpSize;
+	std::cout << "Theoretical Occupancy is " << (double)activeWarps/maxWarps * 100 << "%" << std::endl;
+     }	
+    else{
+	 std::cout << "Invalid warpsize" << std::endl;
+	 std::cout << "No. of Active Blocks per multiprocessor returned by hipOccupancyMaxActiveBlocksPerMultiprocessor is " << numBlock << std::endl;
+         exit(0);
```

Better to return an error and exit only after cleaning up.

---

Not being able to compute the theoretical occupancy should not be a blocking failure. I suggest that we just drop the else part of the code and continue.